reconcile/utils/saasherder.py

```diff
@@ -58,6 +69,87 @@ def __repr__(self):
         return self.__str__()
 
 
+@dataclass
+class TriggerSpecBase:
+    saas_file_name: str
+    env_name: str
+    timeout: Optional[str]
+    pipelines_provider: Optional[dict[str, Any]]
+    resource_template_name: str
+    cluster_name: str
+    namespace_name: str
+    state_content: Any
+
+    @property
+    def state_key(self):
+        raise NotImplementedError("implement this function in inheriting classes")
+
+
+@dataclass
+class TriggerSpecConfig(TriggerSpecBase):
+    reason: Optional[str] = None
+
+    @property
+    def state_key(self):
+        key = (
+            f"{self.saas_file_name}/{self.resource_template_name}/{self.cluster_name}/"
+            + f"{self.namespace_name}/{self.env_name}"
+        )
+        return key
+
+    @staticmethod
+    def from_state_key(
+        key: str,
+        state_content: Any,
+        timeout: Optional[str] = None,
+        pipelines_provider: Optional[dict[str, Any]] = None,
+    ) -> "TriggerSpecConfig":
+        saas_file_name, rt_name, cluster_name, namespace_name, env_name = key.split("/")
+        return TriggerSpecConfig(
+            saas_file_name=saas_file_name,
+            env_name=env_name,
+            timeout=timeout,
+            pipelines_provider=pipelines_provider,
+            resource_template_name=rt_name,
+            cluster_name=cluster_name,
+            namespace_name=namespace_name,
+            state_content=state_content,
+        )
+
+
+@dataclass
+class TriggerSpecMovingCommit(TriggerSpecBase):
+    ref: str
+    reason: Optional[str] = None
+
+    @property
+    def state_key(self):
+        key = (
+            f"{self.saas_file_name}/{self.resource_template_name}/{self.cluster_name}/"
+            + f"{self.namespace_name}/{self.env_name}/{self.ref}"
```

Comment:
nit: i don't think you need the + symbol here. strings in () will be concatenated also without the +

Reply:
https://github.com/app-sre/qontract-reconcile/pull/2782/commits/9ac938b064f4294f42c51eb397b6543f5a2972f1